compiler/rustc_infer/src/infer/relate/type_relating.rs

```diff
@@ -13,9 +14,36 @@ use crate::infer::{DefineOpaqueTypes, InferCtxt, SubregionOrigin};
 
 /// Enforce that `a` is equal to or a subtype of `b`.
 pub struct TypeRelating<'combine, 'a, 'tcx> {
+    // Partially mutable.
```

Comment:
We should mark that it's mutable only because we use it to store the output goals of relating types, which are a pure function of the ty and variance and so this needs no additional tracking.